mobly/snippet/client_base.py

```diff
@@ -403,19 +397,42 @@ def _parse_rpc_response(self, rpc_id, rpc_func_name, response):
             self._device,
             errors.ProtocolError.RESPONSE_MISSING_FIELD % field_name)
 
-    if result['error']:
-      raise errors.ApiError(self._device, result['error'])
     if result['id'] != rpc_id:
       raise errors.ProtocolError(self._device,
                                  errors.ProtocolError.MISMATCHED_API_ID)
-    if result['callback'] is not None:
-      return self.handle_callback(result['callback'], result['result'],
+
+    return result
+
+  def _handle_rpc_response(self, rpc_func_name, response):
+    """Handles the content of RPC response.
+
+    If the RPC response contains error information, it throws an error. If the
+    RPC is asynchronous, it creates and returns a callback handler
+    object. Otherwise, it returns the result field of the response.
+
+    Args:
+      rpc_func_name: str, the name of the snippet function that this RPC
+        triggered on the snippet server.
+      response: dict, the object decoded from the response JSON string.
+
+    Returns:
+      The result of the RPC. If synchronous RPC, it is the result field of the
+      response. If asynchronous RPC, it is the callback handler object.
+
+    Raises:
+      errors.ApiError: if the snippet function executed with errors.
+    """
+
+    if response['error']:
+      raise errors.ApiError(self._device, response['error'])
+    if response['callback'] is not None:
```

Comment:
based on this code, `callback` is not a required field.
but it's listed in `RPC_RESPONSE_REQUIRED_FIELDS`?

Reply:
I think it's more unified to take all existing fields as required and here use json value NULL if no callback. (except the newly added field kwargs)

So I take them as required fields in design, protocol and code. Do we think it's better to change this requirement?

Reply:
In a protocol definition, if you say something is "required", you throw an error if it's absent, because the message is malformed.
Otherwise, it's not required, right?

This is different from whether a value can be null or not.

Reply:
So if you don't think it's necessary to check the existence of `callback` in a response message, `callback` is **not** a required field. It should not be listed in `RPC_RESPONSE_REQUIRED_FIELDS`.

In fact, the first validity check should be making sure everything in `RPC_RESPONSE_REQUIRED_FIELDS` exists in a message.

Reply:
I think we have the same opinion on this code, no?

* First, the callback is a required field in RPC message, so the validity check `_decode_response_string_and_validate_format` checks that all fields are in the message, including `callback`.
* Second, the `_handle_rpc_response` actually reads the value of each field, if the `callback`'s value is null, it takes the response as sync RPC response.

Reply:
I think we have the same opinion on this code, no?

* First, the callback is a required field in RPC message, so the validity check `_decode_response_string_and_validate_format` checks that all fields are in the message, including `callback`.
* Second, the `_handle_rpc_response` actually reads the value of each field, if the `callback`'s value is null, it takes the response as sync RPC response.